lightning/src/routing/network_graph.rs

```diff
@@ -0,0 +1,1640 @@
+	//keep an eye on this to see if the extra memory is a problem
+	pub announcement_message: Option<msgs::NodeAnnouncement>,
+}
+
+impl std::fmt::Display for NodeInfo {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+		write!(f, "features: {}, last_update: {:?}, lowest_inbound_channel_fees: {:?}, channels: {:?}", log_bytes!(self.features.encode()), self.last_update, self.lowest_inbound_channel_fees, &self.channels[..])?;
+		Ok(())
+	}
+}
+
+impl Writeable for NodeInfo {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		(self.channels.len() as u64).write(writer)?;
+		for ref chan in self.channels.iter() {
+			chan.write(writer)?;
+		}
+		self.lowest_inbound_channel_fees.write(writer)?;
+		self.features.write(writer)?;
+		self.last_update.write(writer)?;
+		self.rgb.write(writer)?;
+		self.alias.write(writer)?;
+		(self.addresses.len() as u64).write(writer)?;
+		for ref addr in &self.addresses {
+			addr.write(writer)?;
+		}
+		self.announcement_message.write(writer)?;
+		Ok(())
+	}
+}
+
+const MAX_ALLOC_SIZE: u64 = 64*1024;
+
+impl Readable for NodeInfo {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<NodeInfo, DecodeError> {
+		let channels_count: u64 = Readable::read(reader)?;
+		let mut channels = Vec::with_capacity(cmp::min(channels_count, MAX_ALLOC_SIZE / 8) as usize);
+		for _ in 0..channels_count {
+			channels.push(Readable::read(reader)?);
+		}
+		let lowest_inbound_channel_fees = Readable::read(reader)?;
+		let features = Readable::read(reader)?;
+		let last_update = Readable::read(reader)?;
+		let rgb = Readable::read(reader)?;
+		let alias = Readable::read(reader)?;
+		let addresses_count: u64 = Readable::read(reader)?;
+		let mut addresses = Vec::with_capacity(cmp::min(addresses_count, MAX_ALLOC_SIZE / 40) as usize);
+		for _ in 0..addresses_count {
+			match Readable::read(reader) {
+				Ok(Ok(addr)) => { addresses.push(addr); },
+				Ok(Err(_)) => return Err(DecodeError::InvalidValue),
+				Err(DecodeError::ShortRead) => return Err(DecodeError::BadLengthDescriptor),
+				_ => unreachable!(),
+			}
+		}
+		let announcement_message = Readable::read(reader)?;
+		Ok(NodeInfo {
+			channels,
+			lowest_inbound_channel_fees,
+			features,
+			last_update,
+			rgb,
+			alias,
+			addresses,
+			announcement_message
+		})
+	}
+}
+
+/// Represents the network as nodes and channels between them
+#[derive(PartialEq)]
+pub struct NetworkGraph {
+	channels: BTreeMap<u64, ChannelInfo>,
+	/// A set of known Lightning nodes and their properties
+	nodes: BTreeMap<PublicKey, NodeInfo>,
+}
+
+impl NetworkGraph {
+	/// Returns a list of known valid channels
+	pub fn get_channels<'a>(&'a self) -> &'a BTreeMap<u64, ChannelInfo> { &self.channels }
+	/// Returns a mutable reference to an existing channel
+	pub fn get_channel_mut<'a>(&'a mut self, chan_id: &u64) -> Option<&mut ChannelInfo> { self.channels.get_mut(chan_id) }
```

Comment:
We can't just expose these completely writeable, we want to allow users to add channels/nodes, but only in a way that keeps the internal state consistent. Otherwise we'll end up with panics like the one you showed at https://github.com/rust-bitcoin/rust-lightning/pull/570#discussion_r403817262.